kubernetes_deploy/staging/deployment.yml

```diff
@@ -23,6 +23,20 @@ spec:
           value: https://staging.laalaa.dsd.io
       - image: 926803513772.dkr.ecr.eu-west-1.amazonaws.com/laa-get-access/laa-fala-nginx:master
         name: nginx
+        readinessProbe:
+          httpGet:
+            path: /ping.json
```

Comment:
`/ping.json` is served by the "webapp" container, right?

In my thinking, these containers should check that they are themselves healthy and the pod will check that all containers are healthy.

I think it's not advisable to rely on other containers for readiness and liveness. Could we change that to use a static resource instead, please?

Reply:
💭 We could add a file that serves this purpose, or use the TCP probe instead.

Reply:
https://github.com/ministryofjustice/fala/blob/master/conf/nginx.conf#L23-L25

Nginx is configured to return a 200 for `/ping.json`.

Reply:
> Nginx is configured to return a 200 for `/ping.json`.

Oh, I see, it was added earlier, not part of this pull request.

Could we step back a bit?

- I remember suggesting using `ping.json` in the "nginx" container, but you are right, it's misleading. If someone would call `https://laa-fala-staging.apps..../ping.json`, my expectation would be to ping the Python application as it's expected that everything under `/static` is "nginx" and everything else is "webapp".  

  If we would have `/static/ping.json` the origin would be explicit. What do you think?
- > direct communication to the application isn't possible because communication happens via the uwsgi protocol

  I think _now_ I understand why `socket=:8000` was bothering me. I would always assume each container is viable on its own. If we _need_ "nginx" to try out "webapp" then we would have a tightly coupled dependency and I would not like to force that on anyone.

  Can we change this back to `http=0.0.0.0:8000`, please? I think the general expectation of any web-based component would be to work with HTTP out of the box.
- I can see now that the uwsgi `ping.json` code is unused, I missed that during reviewing #15 😢 We should not have unused code in the codebase. However, if we change the protocol back to `http`, we _could_ use it again, right?

Reply:
> If we would have /static/ping.json the origin would be explicit. What do you think?

Good idea

>I think now I understand why socket=:8000 was bothering me. I would always assume each container is viable on its own. If we need "nginx" to try out "webapp" then we would have a tightly coupled dependency and I would not like to force that on anyone.

>Can we change this back to http=0.0.0.0:8000, please? I think the general expectation of any web-based component would be to work with HTTP out of the box.

You can run the application with by following the readme, with the final command being `python3 ./manage.py runserver` or `DJANGO_SETTINGS_MODULE=fala.settings.local python3 ./manage.py runserver` to use your local settings.

uWSGI is only run when in a Docker container. This is standard for Django applications. 

https://docs.djangoproject.com/en/1.8/howto/deployment/wsgi/

So the set up for local development is:

`the web client <-> Django`

and this set up for docker-compose/Kubernetes, where web server is `nginx`:

`the web client <-> the web server <-> the socket <-> uWSGI <-> Django`

The uWSGI docs says:

>in fact it’s better to use Unix sockets than ports - there’s less overhead.

This is the set up for `cla_public`:

**nginx.conf**
`uwsgi_pass  unix:////tmp/public.sock;`

**cla_public.ini**
`socket = /tmp/public.sock`

>I can see now that the uwsgi ping.json code is unused, I missed that during reviewing #15 😢 We should not have unused code in the codebase. However, if we change the protocol back to http, we could use it again, right?

I'll remove the `ping.json` endpoint from the webapp.